File: Sources/App/Classes/Library/TLOInternetAddressLookup.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2010 - 2018 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "NSObjectHelperPrivate.h"
#import "TPCPreferencesLocal.h"
#import "TLOInternetAddressLookup.h"

NS_ASSUME_NONNULL_BEGIN

#define _requestTimeoutInterval			30.0

@interface TLOInternetAddressLookup ()
@property (nonatomic, weak) id requestDelegate;
@property (nonatomic, strong, nullable) NSURLSession *session;
@property (nonatomic, strong, nullable) NSURLSessionDataTask *connection;
@property (nonatomic, copy, nullable) NSString *address;
@end

@implementation TLOInternetAddressLookup

#pragma mark -
#pragma mark Public API

- (instancetype)init
{
	[self doesNotRecognizeSelector:_cmd];

	return nil;
}

- (instancetype)initWithDelegate:(id <TLOInternetAddressLookupDelegate>)delegate
{
	NSParameterAssert(delegate != nil);

	if ((self = [super init])) {
		self.IPv4AddressIsValid = YES;
		self.IPv6AddressIsValid = YES;

		self.requestDelegate = delegate;

		return self;
	}

	return nil;
}

- (void)performLookup
{
	[self setupConnectionRequest];
}

- (void)cancelLookup
{
	[self _teardownConnectionRequest];
}

- (void)setupConnectionRequest
{
	NSAssert((self.connection == nil),
		@"A lookup is already in progress");

	NSURL *requestURL = [NSURL URLWithString:[self addressSourceURL]];

	NSURLSessionConfiguration *configuration = [NSURLSessionConfiguration ephemeralSessionConfiguration];
	configuration.requestCachePolicy = NSURLRequestReloadIgnoringCacheData;
	configuration.timeoutIntervalForRequest = _requestTimeoutInterval;
	configuration.timeoutIntervalForResource = _requestTimeoutInterval;

	self.session = [NSURLSession sessionWithConfiguration:configuration];

	__weak typeof(self) weakSelf = self;

	self.connection = [self.session dataTaskWithURL:requestURL completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
		dispatch_async(dispatch_get_main_queue(), ^{
			[weakSelf completeLookupWithData:data response:response error:error];
		});
	}];

	[self.connection resume];
}

- (void)_teardownConnectionRequest
{
	[self.connection cancel];
	self.connection = nil;

	[self.session invalidateAndCancel];
	self.session = nil;
}

- (void)teardownConnectionRequest
{
	[self _teardownConnectionRequest];

	[self informDelegate];

	self.address = nil;
}

- (NSString *)addressSourceURL
{
	if ([TPCPreferences fileTransferIPAddressDetectionMethod] == TXFileTransferIPAddressMethodRouterAndThirdParty) {
		return [self thirdPartySourceURL];
	}
	
	return @"https://myip.codeux.com/";
}

- (NSString *)thirdPartySourceURL
{
	NSArray *services = @[
	  @"https://wtfismyip.com/text",
	  @"https://canhazip.com/",
	  @"http://ifconfig.me/ip",
	  @"http://v4.ipv6-test.com/api/myip.php",
	];

	NSUInteger randomIndex = (arc4random() % services.count);

	return services[randomIndex];
}

#pragma mark -
#pragma mark Connection Delegate

- (void)informDelegate
{
	NSString *address = self.address;

	if (address) {
		[self informDelegateLookupReturnedAddress:address];
	} else {
		[self informDelegateLookupFailed];
	}
}

- (void)informDelegateLookupReturnedAddress:(NSString *)address
{
	if ([self.requestDelegate respondsToSelector:@selector(internetAddressLookupReturnedAddress:)]) {
		[self.requestDelegate internetAddressLookupReturnedAddress:address];
	}
}

- (void)informDelegateLookupFailed
{
	if ([self.requestDelegate respondsToSelector:@selector(internetAddressLookupFailed)]) {
		[self.requestDelegate internetAddressLookupFailed];
	}
}

- (void)completeLookupWithData:(NSData *)data response:(NSURLResponse *)response error:(NSError *)error
{
	if (error == nil && [response isKindOfClass:[NSHTTPURLResponse class]] && ((NSHTTPURLResponse *)response).statusCode == 200 && data.length > 0 && data.length <= 1024) {
		NSString *address = [[NSString stringWithData:data encoding:NSUTF8StringEncoding] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];

		if ((address.isIPv4Address && self.IPv4AddressIsValid) ||
			(address.isIPv6Address && self.IPv6AddressIsValid))
		{
			self.address = address;
		}
	} else if (error) {
		LogToConsole("Lookup failed with error: %{public}@", error.localizedDescription);
	}

	[self teardownConnectionRequest];
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/Networking/TLOInternetAddressLookup.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import Foundation
import os

@objc(TLOInternetAddressLookup)
@MainActor
public final class InternetAddressLookup: NSObject {
	private static let requestTimeout: TimeInterval = 30
	private static let firstPartySourceURL = URL(string: "https://api.ipify.org")!
	private static let thirdPartySourceURLs = [
		URL(string: "https://wtfismyip.com/text")!,
		URL(string: "https://canhazip.com/")!,
		URL(string: "https://ifconfig.me/ip")!,
	]
	private static let logger = Logger(
		subsystem: Bundle.main.bundleIdentifier ?? "Glasstual",
		category: "InternetAddressLookup"
	)

	@objc(IPv4AddressIsValid) public var ipv4AddressIsValid = true
	@objc(IPv6AddressIsValid) public var ipv6AddressIsValid = true

	private weak var requestDelegate: TLOInternetAddressLookupDelegate?
	private var session: URLSession?
	private var connection: URLSessionDataTask?
	private var address: String?

	@available(*, unavailable)
	override public init() {
		fatalError("Use init(delegate:)")
	}

	@objc(initWithDelegate:)
	public init(delegate: TLOInternetAddressLookupDelegate) {
		requestDelegate = delegate

		super.init()
	}

	@objc public func performLookup() {
		precondition(connection == nil, "A lookup is already in progress")

		let configuration = URLSessionConfiguration.ephemeral
		configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
		configuration.timeoutIntervalForRequest = Self.requestTimeout
		configuration.timeoutIntervalForResource = Self.requestTimeout

		let session = URLSession(configuration: configuration)
		let sourceURL = addressSourceURL

		self.session = session
		connection = session.dataTask(with: sourceURL) { [weak self] data, response, error in
			Task { @MainActor [weak self] in
				self?.completeLookup(data: data, response: response, error: error)
			}
		}
		connection?.resume()
	}

	@objc public func cancelLookup() {
		teardownConnection()
	}

	@objc(addressFromData:response:allowIPv4:allowIPv6:)
	public nonisolated static func address(
		from data: Data?,
		response: URLResponse?,
		allowIPv4: Bool,
		allowIPv6: Bool
	) -> String? {
		guard
			let response = response as? HTTPURLResponse,
			response.statusCode == 200,
			let data,
			!data.isEmpty,
			data.count <= 1024,
			let address = String(data: data, encoding: .utf8)?
			.trimmingCharacters(in: .whitespacesAndNewlines)
		else {
			return nil
		}

		let bridgedAddress = address as NSString

		guard
			(allowIPv4 && bridgedAddress.isIPv4Address)
			|| (allowIPv6 && bridgedAddress.isIPv6Address)
		else {
			return nil
		}

		return address
	}

	private var addressSourceURL: URL {
		if TPCPreferences.fileTransferIPAddressDetectionMethod() == .routerAndThirdParty {
			return Self.thirdPartySourceURLs.randomElement()!
		}

		return Self.firstPartySourceURL
	}

	private func completeLookup(data: Data?, response: URLResponse?, error: Error?) {
		if let error {
			Self.logger.error("Lookup failed: \(error.localizedDescription, privacy: .public)")
		} else {
			address = Self.address(
				from: data,
				response: response,
				allowIPv4: ipv4AddressIsValid,
				allowIPv6: ipv6AddressIsValid
			)
		}

		teardownConnection()
		informDelegate()
		address = nil
	}

	private func teardownConnection() {
		connection?.cancel()
		connection = nil

		session?.invalidateAndCancel()
		session = nil
	}

	private func informDelegate() {
		if let address {
			requestDelegate?.internetAddressLookupReturnedAddress(address)
		} else {
			requestDelegate?.internetAddressLookupFailed()
		}
	}
}
